registry/registry.go

```diff
@@ -759,24 +836,297 @@ func downloadVersion(opts *VersionOptions) (ver *Version, attachments []*kivik.A
+	// local registry url for future downloads
+	ver.URL = opts.RegistryURL.String()
+	ver.Sha256 = opts.Sha256
+	ver.Editor = parsedManifest.Editor
+	ver.Manifest = manifestContent
+	ver.Size = tarball.Size
+	ver.TarPrefix = tarball.TarPrefix
+	ver.CreatedAt = time.Now().UTC()
+	return
+}
+
+func getIconPath(parsedManifest *Manifest, opts *VersionOptions) string {
+	var iconPath string
+	if opts.Icon != "" {
+		iconPath = opts.Icon
+	} else {
+		iconPath = parsedManifest.Icon
+	}
+	if iconPath != "" {
+		iconPath = path.Join("/", iconPath)
+	}
+	return iconPath
+}
+
+func getPartnershipIconPath(parsedManifest *Manifest, opts *VersionOptions) string {
+	var partnershipIconPath string
+	if opts.Partnership.Icon != "" {
+		partnershipIconPath = opts.Partnership.Icon
+	} else {
+		partnershipIconPath = parsedManifest.Partnership.Icon
+	}
+	if partnershipIconPath != "" {
+		partnershipIconPath = path.Join("/", partnershipIconPath)
+	}
+	return partnershipIconPath
+}
+
+func getScreenshotPaths(parsedManifest *Manifest, opts *VersionOptions) []string {
+	var screenshotPaths []string
+	if opts.Screenshots != nil {
+		screenshotPaths = opts.Screenshots
+		for i, shot := range screenshotPaths {
+			screenshotPaths[i] = path.Join("/", shot)
+		}
+	} else {
+		for _, shot := range parsedManifest.Screenshots {
+			screenshotPaths = append(screenshotPaths, path.Join("/", shot))
+		}
+		for _, locale := range parsedManifest.Locales {
+			for _, shot := range locale.Screenshots {
+				shot = path.Join("/", shot)
+				if !stringInArray(shot, screenshotPaths) {
+					screenshotPaths = append(screenshotPaths, shot)
+				}
+			}
+		}
+	}
+
+	return screenshotPaths
+}
+
+// getAssetFilename computes the asset filename to write to the FS (icon,
+// partnership_icon, screenshot, ...)
+func getAssetFilename(iconPath, partnershipIconPath, name string, screenshotPaths []string) string {
+	isIcon := iconPath != "" && name == iconPath
+	isPartnershipIcon := partnershipIconPath != "" && name == partnershipIconPath
+
+	isShot := !isIcon && stringInArray(name, screenshotPaths)
+	if !isIcon && !isPartnershipIcon && !isShot {
+		return ""
+	}
+
+	// Sets filename
+	var filename string
+	if isIcon {
+		filename = "icon"
+	} else if isShot {
+		filename = name
+	} else if isPartnershipIcon {
+		filename = "partnership_icon"
+	} else {
+		panic("unreachable")
+	}
+
+	return filename
+}
+
+// HandleAssets handles all the assets of the app tarball (icon, partnership
+// icon, screenshots). Appened to attachments
+func HandleAssets(tarball *Tarball, opts *VersionOptions) ([]*kivik.Attachment, error) {
+	var attachments []*kivik.Attachment
+	parsedManifest := tarball.Manifest
+
+	iconPath := getIconPath(parsedManifest, opts)
+	partnershipIconPath := getPartnershipIconPath(parsedManifest, opts)
+	screenshotPaths := getScreenshotPaths(parsedManifest, opts)
+
+	// Re-reading tarball content for assets
+	if len(screenshotPaths) > 0 || iconPath != "" || partnershipIconPath != "" {
```

Comment:
[nit] I'd prefer an explicit return here instead of a long `if`:

```go
if len(screenshotPaths) == 0 && iconPath == "" && partnershipIconPath == "" {
    return
}
```